test/unit_tests/24_dma_endBlock/test.cpp

```diff
@@ -0,0 +1,107 @@
+#include <xaiengine.h>
+#include "test_library.h"
+
+#define LOCK_TIMEOUT 100
+#define HIGH_ADDR(addr)	((addr & 0xffffffff00000000) >> 32)
+#define LOW_ADDR(addr)	(addr & 0x00000000ffffffff)
+#define MLIR_STACK_OFFSET 4096
+
+#include "aie_inc.cpp"
+
+int
+main(int argc, char *argv[])
+{
+    printf("DMA_endBlock test start.\n");
+
+    // Standard helper function for initializing and configuring AIE array.
+    // The host is used to initialize/ configure/ program the AIE array.
+    // ------------------------------------------------------------------------
+    // aie_libxaie_ctx_t - AIE config struct
+    // mlir_aie_init_device ##### TODO #######
+    // mlir_aie_configure_cores - Reset cores and locks. Load elfs.
+    // mlir_aie_configure_switchboxes - Switchboxes not used in this example.
+    // mlir_aie_configure_dmas - TileDMAs not used in this example.
+    // mlir_aie_initialize_locks - Locks not used in this example.
+    aie_libxaie_ctx_t *_xaie = mlir_aie_init_libxaie();
+    mlir_aie_init_device(_xaie);                        
+    mlir_aie_configure_cores(_xaie);
+    mlir_aie_configure_switchboxes(_xaie);
+    mlir_aie_configure_dmas(_xaie);
+    mlir_aie_initialize_locks(_xaie);
+
+    int errors = 0;
+
+    // Helper function to clear tile data memory
+    mlir_aie_clear_tile_memory(_xaie, 1, 4);
+    mlir_aie_clear_tile_memory(_xaie, 3, 4);
+
+    // Check the buffer value at index 3 to ensure it is zeroed out
+    // prior to running our simple kernel.
+    // ------------------------------------------------------------------------
+    // mlir_aie_read_buffer_a14 - helper function to read tile local
+    // memory at an offset (offset=3 in this case). _a14 maps to the 
+    // symbolic buffer name defined in aie.mlir.
+    //
+    // mlir_aie_check - helper function to compare values to expected
+    // golden value and print error message to stdout and increment 
+    // "errors" variable if mismatch occurs.
+    mlir_aie_check("Before start cores:", mlir_aie_read_buffer_a14(_xaie, 3), 0,
+                   errors);
+    mlir_aie_check("Before start cores:", mlir_aie_read_buffer_a34(_xaie, 5), 0,
+                   errors);
+
+    // Helper function to enable all AIE cores
+    printf("Start cores\n");
+    mlir_aie_start_cores(_xaie);
+
+    mlir_aie_acquire_lock_a14_6(_xaie, 0, 0);
+    mlir_aie_acquire_lock_a34_7(_xaie, 1, 0);
+
+    mlir_aie_release_lock_a14_6(_xaie, 1, 0);
+    mlir_aie_release_lock_a34_7(_xaie, 0, 0);
+
+    // Wait time for cores to run. Number used here is much larger than needed.
+    usleep(100);
+
+    // Check buffer at index 3 again for expected value of 14 for tile(1,4)    
+    mlir_aie_check("After start cores:", mlir_aie_read_buffer_a14(_xaie, 3), 14,
+                   errors);
+    // Check buffer at index 5 again for expected value of 114 for tile(3,4)    
+    mlir_aie_check("After start cores:", mlir_aie_read_buffer_a34(_xaie, 5), 114,
+                   errors);
```

Comment:
It's not clear to me what this is testing, exactly.  How does this fail without the patch?

Reply:
Without the patch, the cf.br ^end becomes and cf.br ^bd0. This means that the test.cpp calls to the lock functions will cause the DMAs to run again thereby overwriting the output with a 0. The test will fail if that's the case.